gcs_contents_manager.py

```diff
@@ -428,3 +449,262 @@ def rename_file(self, old_path, new_path):
+        child_obj['path'] = path_prefix
+        child_obj['name'] = path_prefix
+        child_obj['writable'] = False
+        contents.append(child_obj)
+      dir_obj['content'] = contents
+      dir_obj['created'] = contents[0]['created']
+      dir_obj['last_modified'] = contents[0]['last_modified']
+      return dir_obj
+    try:
+      cm, relative_path, path_prefix = self._content_manager(path)
+      if not cm:
+        raise HTTPError(404, 'Not Found')
+
+      model = cm.get(relative_path, content=content, type=type, format=format)
+      self._make_model_relative(model, path_prefix)
+      return model
+    except HTTPError as err:
+      raise err
+    except Exception as ex:
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(type(ex), str(ex)))
+
+  def save(self, model, path):
+    if path in ['', '/']:
+      raise HTTPError(403, 'The top-level directory is read-only')
+    try:
+      self.run_pre_save_hook(model=model, path=path)
+
+      cm, relative_path, path_prefix = self._content_manager(path)
+      if not cm:
+        raise HTTPError(404, 'Not Found')
+      if relative_path in ['', '/']:
+        raise HTTPError(403, 'The top-level directory contents are read-only')
+
+      if 'path' in model:
+        model['path'] = relative_path
+
+      model = cm.save(model, relative_path)
+      if 'path' in model:
+        model['path'] = path
+      return model
+    except HTTPError as err:
+      raise err
+    except Exception as ex:
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(type(ex), str(ex)))
+
+  def delete_file(self, path):
+    if path in ['', '/']:
+      raise HTTPError(403, 'The top-level directory is read-only')
+    try:
+      cm, relative_path, path_prefix = self._content_manager(path)
+      if not cm:
+        raise HTTPError(404, 'Not Found')
+      if relative_path in ['', '/']:
+        raise HTTPError(403, 'The top-level directory contents are read-only')
+      return cm.delete_file(relative_path)
+    except OSError as err:
+      # The built-in file contents manager will not attempt to wrap permissions
+      # errors when deleting files if they occur while trying to move the
+      # to-be-deleted file to the trash, because the underlying send2trash
+      # library does not set the errno attribute of the raised OSError.
+      #
+      # To work around this we explicitly catch such errors, check if they
+      # start with the magic text "Permission denied", and then wrap them
+      # in an HTTPError.
+      if str(err).startswith('Permission denied'):
+        raise HTTPError(403, str(err))
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(err.errno, str(err)))
+    except HTTPError as err:
+      raise err
+    except Exception as ex:
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(type(ex), str(ex)))
+
+  def rename_file(self, old_path, new_path):
+    if old_path in ['', '/']:
```

Comment:
I think this check needs to happen on new_path as well so that you don't return 404 when trying to rename to root. (Or at least I'm not sure which is the intended behavior)

Reply:
Done